tensorflow/python/ops/array_ops.py

```diff
@@ -575,6 +579,8 @@ def _slice_helper(tensor, slice_spec, var=None):
     TypeError: If the slice indices aren't int, slice, ellipsis,
       tf.newaxis or scalar int32/int64 tensors.
   """
+  if isinstance(slice_spec, (np.ndarray, ops.Tensor)):
```

Comment:
`tensor[tf.constant(0)]` should still index out the 0th element from the first axis --- please add a test if there isn't one already

Reply:
@shoyer, the current code would report for tensor[tf.constant(0)]: ValueError: mask cannot be scalar.
That is the reason why my previous code checked for both np.ndarray and ops.Tensor if the input is scalar, if so then the example such as tensor[tf.constant(0)] will be treated as the original index slicing operation.

Reply:
> the current code would report for tensor[tf.constant(0)]: ValueError: mask cannot be scalar.

Right, so this needs to be changed, e.g., to explicitly check dtype instead.

We don't want to break any existing uses of indexing. This should only work in cases that always resulted in an error previously.